crates/walrus-sdk/src/client.rs

```diff
@@ -441,6 +510,172 @@ impl<T: ReadClient> Client<T> {
+                        "Timeout reached after {:?} while retrieving slivers",
+                        timeout_duration
+                    );
+                    break;
+                }
+            }
+            if let Some(max_attempts) = max_attempts {
+                if attempts > max_attempts {
+                    tracing::debug!("Max attempts ({}) reached", max_attempts);
+                    break;
+                }
+            }
+
+            tracing::debug!("Retrieving slivers: {:?}", sliver_selector);
+            match self
+                .retrieve_slivers(metadata, &sliver_selector, certified_epoch)
+                .await
+            {
+                Ok(new_slivers) => {
+                    // Track which indices we've successfully retrieved.
+                    for sliver in new_slivers {
+                        sliver_selector.complete_sliver(&sliver.index);
+                        all_slivers.insert(sliver.index, sliver);
+                    }
+                }
+                Err(e) => {
+                    // TODO(heliu): if the error is not retriable, return the error immediately.
+                    tracing::warn!("Error retrieving slivers: {:?}", e);
+                    last_error = Some(e);
+                }
+            }
+
+            attempts += 1;
+            if !sliver_selector.is_empty() {
+                tokio::time::sleep(Duration::from_millis(retry_delay_ms)).await;
+            }
+        }
+
+        if all_slivers.len() != sliver_indices.len() {
+            Err(ClientError::from(ClientErrorKind::Other(
+                format!("Failed to retrieve slivers: {:?}", last_error).into(),
+            )))
+        } else {
+            Ok(all_slivers.into_values().collect())
+        }
+    }
+
+    /// Retrieves specific slivers from storage nodes based on their indices.
+    #[tracing::instrument(level = Level::ERROR, skip_all)]
+    async fn retrieve_slivers<'a, E: EncodingAxis>(
+        &self,
+        metadata: &'a VerifiedBlobMetadataWithId,
+        sliver_selector: &SliverSelector<E>,
+        certified_epoch: Epoch,
+    ) -> ClientResult<Vec<SliverData<E>>>
+    where
+        SliverData<E>: TryFrom<Sliver>,
+    {
+        let blob_id = metadata.blob_id();
+        tracing::info!("starting to retrieve slivers {:?}", sliver_selector);
+        self.check_blob_id(blob_id)?;
+
+        // Create a progress bar to track the progress of the sliver retrieval.
+        let progress_bar: indicatif::ProgressBar =
+            styled_progress_bar(sliver_selector.num_slivers() as u64);
+        progress_bar.set_message(format!(
+            "requesting {} slivers",
+            sliver_selector.num_slivers()
+        ));
+
+        let committees = self.get_committees().await?;
+        let comms = self
+            .communication_factory
+            .node_read_communications(&committees, certified_epoch)?;
+
+        // Create requests to get all slivers from all nodes.
```

Comment:
This comment is not correct:
```suggestion
        // Create requests to get all requested slivers.
```